### ISSUE TITLE
[BUG] - Unstable DNS plan output wants to create existing records

### ISSUE THREAD

**Issue by jdmulloy** [vultr/terraform-provider-vultr#86]
**Describe the bug**
After painstakingly re-importing my records after upgrading the provider to the new version which uses the v2 API with UUID IDs when I run "terraform plan" I keep getting different outputs to records it wants to create that already exist in the API and in the terraform state. It seems to be randomly not finding a few of my DNS records each time. This seems to be happening more on my main domain which has about 217 records.

**To Reproduce**
1. Run terraform plan multiple times and see that the output is different each time
```
------------------------------------------------------------------------                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                               
An execution plan has been generated and is shown below.                                                                                                                                                                                                                       
Resource actions are indicated with the following symbols:                                                                                                                                                                                                                     
  + create                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                               
Terraform will perform the following actions:                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                               
  # vultr_dns_record.AAAA_joeandsarah_web_vps-vu-fl-1a_jdmulloy_com will be created                                                                                                                                                                                            
  + resource "vultr_dns_record" "AAAA_joeandsarah_web_vps-vu-fl-1a_jdmulloy_com" {                                                                                                                                                                                             
      + data   = "2001:19f0:9002:451::4"                                                                                                                                                                                                                                       
      + domain = "jdmulloy.com"                                                                                                                                                                                                                                                
      + id     = (known after apply)                                                                                                                                                                                                                                           
      + name   = "joeandsarah.web.vps-vu-fl-1a"                                                                                                                                                                                                                                
      + ttl    = 120                                                                                                                                                                                                                                                           
      + type   = "AAAA"                                                                                                                                                                                                                                                        
    }                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                               
  # vultr_dns_record.AAAA_vps-vu-fl-1a_jdmulloy_com will be created                                                                                                                                                                                                            
  + resource "vultr_dns_record" "AAAA_vps-vu-fl-1a_jdmulloy_com" {                                                                                                                                                                                                             
      + data   = "2001:19f0:9002:451::1"                                                                                                                                                                                                                                       
      + domain = "jdmulloy.com"                                                                                                                                                                                                                                                
      + id     = (known after apply)                                                                                                                                                                                                                                           
      + name   = "vps-vu-fl-1a"                                                                                                                                                                                                                                                
      + ttl    = 120                                                                                                                                                                                                                                                           
      + type   = "AAAA"                                                                                                                                                                                                                                                        
    }                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                               
  # vultr_dns_record.CNAME_ipv6_test_joeandsarah_web_fl_jdmulloy_com will be created                                                                                                                                                                                           
  + resource "vultr_dns_record" "CNAME_ipv6_test_joeandsarah_web_fl_jdmulloy_com" {                                                                                                                                                                                            
      + data   = "ipv6.web.fl.jdmulloy.com"                                                                                                                                                                                                                                    
      + domain = "jdmulloy.com"                                                                                                                                                                                                                                                
      + id     = (known after apply)                                                                                                                                                                                                                                           
      + name   = "ipv6.test.joeandsarah.web.fl"                                                                                                                                                                                                                                
      + ttl    = 120                                                                                                                                                                                                                                                           
      + type   = "CNAME"                                                                                                                                                                                                                                                       
    }                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                               
Plan: 3 to add, 0 to change, 0 to destroy.                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                               
------------------------------------------------------------------------
```

```
------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # vultr_dns_record.AAAA_joeandsarah_web_vps-vu-fl-1a_jdmulloy_com will be created
  + resource "vultr_dns_record" "AAAA_joeandsarah_web_vps-vu-fl-1a_jdmulloy_com" {
      + data   = "2001:19f0:9002:451::4"
      + domain = "jdmulloy.com"
      + id     = (known after apply)
      + name   = "joeandsarah.web.vps-vu-fl-1a"
      + ttl    = 120
      + type   = "AAAA"
    }

  # vultr_dns_record.A_ip-web_vps-vu-fl-1a_jdmulloy_com will be created
  + resource "vultr_dns_record" "A_ip-web_vps-vu-fl-1a_jdmulloy_com" {
      + data   = "45.77.197.0"
      + domain = "jdmulloy.com"
      + id     = (known after apply)
      + name   = "ip-web.vps-vu-fl-1a"
      + ttl    = 120
      + type   = "A"
    }

  # vultr_dns_record.CAA_ipv4_mx1_jdmulloy_com will be created
  + resource "vultr_dns_record" "CAA_ipv4_mx1_jdmulloy_com" {
      + data   = "0 issue \"letsencrypt.org\""
      + domain = "jdmulloy.com"
      + id     = (known after apply)
      + name   = "ipv4.mx1"
      + ttl    = 3600
      + type   = "CAA"
    }

  # vultr_dns_record.CAA_jdmulloy_net will be created
  + resource "vultr_dns_record" "CAA_jdmulloy_net" {
      + data   = "0 issue \"letsencrypt.org\""
      + domain = "jdmulloy.net"
      + id     = (known after apply)
      + ttl    = 3600
      + type   = "CAA"
    }

  # vultr_dns_record.CAA_joemulloy_com will be created
  + resource "vultr_dns_record" "CAA_joemulloy_com" {
      + data   = "0 issue \"letsencrypt.org\""
      + domain = "joemulloy.com"
      + id     = (known after apply)
      + ttl    = 3600
      + type   = "CAA"
    }

  # vultr_dns_record.CNAME_ipv6_jdmip_net will be created
  + resource "vultr_dns_record" "CNAME_ipv6_jdmip_net" {
      + data   = "ip6.jdmip.net"
      + domain = "jdmip.net"
      + id     = (known after apply)
      + name   = "ipv6"
      + ttl    = 3600
      + type   = "CNAME"
    }

Plan: 6 to add, 0 to change, 0 to destroy.

------------------------------------------------------------------------
```

**Expected behavior**
Terraform plan shows no changes

**Desktop (please complete the following information where applicable:**
 - OS: FreeBSD 12.1

**Comment by ddymko** [comment]
@jdmulloy thanks for creating the ticket - we'll look into this

**Comment by ddymko** [comment]
@jdmulloy I can't replicate the records wanting to be recreated after an import.

I created a domain + record on v1.5.0 then upgraded to v2.0.0 and deleted the domain + record from my statefile and reimported them and finished writing the resource.

```
➜  tf-playground terraform import vultr_dns_record.rec testdymko.org,82016494-453f-49b7-9bd5-d4a2398724e8
vultr_dns_record.rec: Importing from ID "testdymko.org,82016494-453f-49b7-9bd5-d4a2398724e8"...
vultr_dns_record.rec: Import prepared!
  Prepared vultr_dns_record for import
vultr_dns_record.rec: Refreshing state... [id=82016494-453f-49b7-9bd5-d4a2398724e8]

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.

➜  tf-playground tp

Error: Missing required argument

  on main.tf line 30, in resource "vultr_dns_record" "rec":
  30: resource "vultr_dns_record" "rec" {

The argument "name" is required, but no definition was found.


Error: Missing required argument

  on main.tf line 30, in resource "vultr_dns_record" "rec":
  30: resource "vultr_dns_record" "rec" {

The argument "type" is required, but no definition was found.


Error: Missing required argument

  on main.tf line 30, in resource "vultr_dns_record" "rec":
  30: resource "vultr_dns_record" "rec" {

The argument "data" is required, but no definition was found.


Error: Missing required argument

  on main.tf line 30, in resource "vultr_dns_record" "rec":
  30: resource "vultr_dns_record" "rec" {

The argument "domain" is required, but no definition was found.
```
Once I added these to my resource in main.tf and reran `terraform plan` the resources matched 

![Screen Shot 2020-12-07 at 3 37 49 PM](https://user-images.githubusercontent.com/1588386/101403367-24040500-38a3-11eb-8847-e6390616306f.png)

We may need to get log output from your TF runs to see why these calls aren't finding your records this can be done by setting `TF_LOG=TRACE` in your shell and rerunning the plan. 

Also just out of curiosity could you share the commands you used to import the DNS records?

**Comment by jdmulloy** [comment]
I used commands like the following to import records. In most cases I also had to ```terraform state rm``` the existing state for the resource.

```
terraform import vultr_dns_record.A_mx1_jdmulloy_com jdmulloy.com,2ab7a3f6-7fe5-42a9-8a95-a3d357e53f92
```

Versions
```
# terraform version
Terraform v0.13.4
+ provider registry.terraform.io/-/vultr v2.1.0
+ provider registry.terraform.io/vultr/vultr v2.1.0

Your version of Terraform is out of date! The latest version
is 0.14.0. You can update by downloading from https://www.terraform.io/downloads.html
```

Plan and Trace output
https://gist.github.com/jdmulloy/e681ba24d3f58e67dd567b16f627fb6f

**Comment by Oogy** [comment]
@jdmulloy Thank you for the logs - could you please confirm the `vultr_dns_record.AAAA_joeandsarah_web_vps-vu-fl-1a_jdmulloy_com` resource has been imported? The logs seem to indicate while it is present in your terraform configuration, it _has not_ been imported:
```
2020/12/08 15:15:11 [DEBUG] Resource instance state not found for node "vultr_dns_record.AAAA_joeandsarah_web_vps-vu-fl-1a_jdmulloy_com", instance vultr_dns_record.AAAA_joeandsarah_web_vps-vu-fl-1a_jdmulloy_com
[...]
2020/12/08 15:15:11 [TRACE] EvalReadState: no state present for vultr_dns_record.AAAA_joeandsarah_web_vps-vu-fl-1a_jdmulloy_com
```

**Comment by jdmulloy** [comment]
@Oogy Turns out that one wasn't in the state. Importing it fixed it, but now there are some that are in the state that it wants to create that also already exist.

```
------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # vultr_dns_record.AAAA_ip-web_vps-vu-nj-1a_jdmulloy_com will be created
  + resource "vultr_dns_record" "AAAA_ip-web_vps-vu-nj-1a_jdmulloy_com" {
      + data   = "2001:19f0:5:331::4"
      + domain = "jdmulloy.com"
      + id     = (known after apply)
      + name   = "ip-web.vps-vu-nj-1a"
      + ttl    = 120
      + type   = "AAAA"
    }

  # vultr_dns_record.A_ipv4_joeandsarah_web_vps-vu-fl-1a_jdmulloy_com will be created
  + resource "vultr_dns_record" "A_ipv4_joeandsarah_web_vps-vu-fl-1a_jdmulloy_com" {
      + data   = "45.77.197.0"
      + domain = "jdmulloy.com"
      + id     = (known after apply)
      + name   = "ipv4.joeandsarah.web.vps-vu-fl-1a"
      + ttl    = 120
      + type   = "A"
    }

  # vultr_dns_record.CAA_drive_jdmulloy_com will be created
  + resource "vultr_dns_record" "CAA_drive_jdmulloy_com" {
      + data   = "0 issue \"letsencrypt.org\""
      + domain = "jdmulloy.com"
      + id     = (known after apply)
      + name   = "drive"
      + ttl    = 3600
      + type   = "CAA"
    }

  # vultr_dns_record.CNAME_personal_gmail_jdmulloy_com will be created
  + resource "vultr_dns_record" "CNAME_personal_gmail_jdmulloy_com" {
      + data   = "jdmulloy.com"
      + domain = "jdmulloy.com"
      + id     = (known after apply)
      + name   = "personal.gmail"
      + ttl    = 3600
      + type   = "CNAME"
    }

Plan: 4 to add, 0 to change, 0 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.

root@acme:/terraform/jdmulloy-vultr-terraform/dns_static # grep AAAA_ip-web_vps-vu-nj-1a_jdmulloy_com terraform.tfstate
      "name": "AAAA_ip-web_vps-vu-nj-1a_jdmulloy_com",
root@acme:/terraform/jdmulloy-vultr-terraform/dns_static # grep ip-web.vps-vu-nj-1a jdm
jdmip.txt         jdmulloy.com.txt  jdmulloy.txt      
root@acme:/terraform/jdmulloy-vultr-terraform/dns_static # grep ip-web.vps-vu-nj-1a uuids_jdmulloy.com.txt 
3e92ce0b-c47e-4c0b-bf23-9f47a73ea8f5    A       ip-web.vps-vu-nj-1a     45.63.8.239                                                                                                                                                                                           0120
9876654a-5bea-42ed-9365-4c697824c609    CAA     ip-web.vps-vu-nj-1a                     0 issue "letsencrypt.org"       0               3600
276e5305-c22e-414e-b0f0-9a52fe223ab8    AAAA    ip-web.vps-vu-nj-1a     2001:19f0:5:331::4      0               120
root@acme:/terraform/jdmulloy-vultr-terraform/dns_static # grep 276e5305-c22e-414e-b0f0-9a52fe223ab8 terraform.tfstate
            "id": "276e5305-c22e-414e-b0f0-9a52fe223ab8",
root@acme:/terraform/jdmulloy-vultr-terraform/dns_static #
```

**Comment by jdmulloy** [comment]
I ran plan again with trace and I checked that the state does exist, but it can't find it.

https://gist.github.com/jdmulloy/75cf902cabef08f37aec61229fa6fdc3

**Comment by ddymko** [comment]
@jdmulloy 

Looking at the four records you provided the 1st once has a mismatched TTL

```
 + resource "vultr_dns_record" "AAAA_ip-web_vps-vu-nj-1a_jdmulloy_com" {
      + data   = "2001:19f0:5:331::4"
      + domain = "jdmulloy.com"
      + id     = (known after apply)
      + name   = "ip-web.vps-vu-nj-1a"
      + ttl    = 120
      + type   = "AAAA"
    }
```

The API returns 

```
2020-12-08T20:14:43.657Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: ---[ REQUEST ]---------------------------------------
2020-12-08T20:14:43.657Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: GET /v2/domains/jdmulloy.com/records/0d12f216-7d3e-44e8-92d4-b7b442c05c4b HTTP/1.1
2020-12-08T20:14:43.657Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: Host: api.vultr.com
2020-12-08T20:14:43.657Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: User-Agent: Terraform/1.8.0
2020-12-08T20:14:43.657Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: Accept: application/json
2020-12-08T20:14:43.657Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: Content-Type: application/json
2020-12-08T20:14:43.657Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: Accept-Encoding: gzip
2020-12-08T20:14:43.657Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: 
2020-12-08T20:14:43.657Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: 
2020-12-08T20:14:43.657Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: -----------------------------------------------------
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: 2020/12/08 20:14:43 [DEBUG] Vultr API Response Details:
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: ---[ RESPONSE ]--------------------------------------
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: HTTP/2.0 200 OK
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: Cache-Control: no-cache
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: Content-Type: application/json
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: Date: Tue, 08 Dec 2020 20:14:43 GMT
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: Expires: Tue, 08 Dec 2020 20:14:42 GMT
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: Server: nginx
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: Strict-Transport-Security: max-age=31536000
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: X-Content-Type-Options: nosniff
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: X-Frame-Options: DENY
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: X-Robots-Tag: noindex,noarchive
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: 
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: {
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0:  "record": {
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0:   "id": "3e92ce0b-c47e-4c0b-bf23-9f47a73ea8f5",
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0:   "type": "A",
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0:   "name": "ip-web.vps-vu-nj-1a",
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0:   "data": "45.63.8.239",
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0:   "priority": 0,
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0:   "ttl": 120
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0:  }
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: }
2020-12-08T20:14:43.702Z [DEBUG] plugin.terraform-provider-vultr_v2.1.0: -----------------------------------------------------
```

As for the other three

```
  # vultr_dns_record.A_ipv4_joeandsarah_web_vps-vu-fl-1a_jdmulloy_com will be created
  # vultr_dns_record.CAA_drive_jdmulloy_com will be created
  # vultr_dns_record.CNAME_personal_gmail_jdmulloy_com will be created
```

They do show up in the debug logs you provided they all have have proper `200 OK` response with the right data. The other thing I see in the debug is the that the rate limit is being hit. In v1 it was 3r/s in v2 it was updated to 20r/s. Granted this isn't a huge leap the amount of records and other reads you may possible have can cause to hit the rate limit.

I would suggest first increasing the rate limit this may account for some randomness.  You can update your provider section with the following.  

```
provider "vultr" {
  api_key = "VULTR_API_KEY"
  rate_limit = 1300
  retry_limit = 6
}
```

This will help us narrow down what is causing the randomness/issues here.

**Comment by jdmulloy** [comment]
@ddymko The rate limit thing does seem to be the culprit. Increasing the limit fixes it. Also I tried the way I used to run it with the v1 API with parallelism=1 and that works too. I had to run it that way on v1 or else it would crash due to the rate limit. It seemed like the rate limit setting for v1 at least was per thread, so even increasing it didn't help.

So it seems slowing it down fixes my issue. Thanks. Although I'm happy to help give more debugging info if you want to make this work out of the box.

```
terraform plan -parallelism=1
```
